wgpu-hal/src/auxil/dxgi/result.rs

```diff
@@ -1,37 +1,46 @@
 use std::borrow::Cow;
 
-use winapi::shared::winerror;
+use windows::Win32::{Foundation, Graphics::Dxgi};
 
 pub(crate) trait HResult<O> {
     fn into_result(self) -> Result<O, Cow<'static, str>>;
     fn into_device_result(self, description: &str) -> Result<O, crate::DeviceError>;
 }
-impl HResult<()> for i32 {
-    fn into_result(self) -> Result<(), Cow<'static, str>> {
-        if self >= 0 {
-            return Ok(());
-        }
+impl<T> HResult<T> for windows::core::Result<T> {
+    fn into_result(self) -> Result<T, Cow<'static, str>> {
+        // TODO: use windows-rs built-in error formatting?
         let description = match self {
-            winerror::E_UNEXPECTED => "unexpected",
-            winerror::E_NOTIMPL => "not implemented",
-            winerror::E_OUTOFMEMORY => "out of memory",
-            winerror::E_INVALIDARG => "invalid argument",
-            _ => return Err(Cow::Owned(format!("0x{:X}", self as u32))),
+            Ok(t) => return Ok(t),
+            Err(e) if e.code() == Foundation::E_UNEXPECTED => "unexpected",
+            Err(e) if e.code() == Foundation::E_NOTIMPL => "not implemented",
+            Err(e) if e.code() == Foundation::E_OUTOFMEMORY => "out of memory",
+            Err(e) if e.code() == Foundation::E_INVALIDARG => "invalid argument",
+            Err(e) => return Err(Cow::Owned(format!("{e:?}"))),
         };
         Err(Cow::Borrowed(description))
     }
-    fn into_device_result(self, description: &str) -> Result<(), crate::DeviceError> {
+    fn into_device_result(self, description: &str) -> Result<T, crate::DeviceError> {
         #![allow(unreachable_code)]
 
+        let err_code = if let Err(err) = &self {
+            Some(err.code())
+        } else {
+            None
+        };
```

Comment:
... And with that, getting rid of this ownership-workaround.

Reply:
(same as above)